keps/sig-instrumentation/20190605-metrics-validation-and-verification.md

```diff
@@ -0,0 +1,214 @@
+Sig-instrumentation will own a directory under `test/instrumentation`.
+There will be a subdirectory `testdata` in which a file `stable-metrics-list.txt` will live.
+This file will be owned by sig-instrumentation.
+Metrics conformance tests will involve a static analysis script which will traverse the entire codebase and look for
+metrics which are annotated as 'stable'.
+For each stable metrics, this script will generate a stringified version of metric metadata (i.e. name, type, labels)
+which will then be appended together in lexographic order. This will be the output of this script.
+
+We will add a pre-submit check, which will run in our CI pipeline, which will run our script with the current changes
+and compare that to existing, committed file. If there is a difference, the pre-submit check will fail.
+In order to pass the pre-submit check, the original submitter of the PR will have to run a script `
+test/instrumentation/update-stable-metrics.sh` which will run our static analysis code and overwrite `stable-metrics-list.yaml`.
+This will cause `sig-instrumentation` to be tagged for approvals on the PR (since they own that file).
+ 
+ 
+### Static Analysis
+ 
+Similarly to conformance test proposed analysis will be performed on golang source code using default abstract syntax tree parser `go/ast`.
+ Handling all possible cases for how metrics can be instantiated would require executing the code itself and is not practical.
+ To reduce number of cases needed handled we will be making assumption of non malicious intent of contributors.
+ There are just too many ways one golang struct can be instantiated that would hide potential stable metrics.
+ To ensure correctness and code simplicity we will be restricting how stable metrics can be declared to one format described below.
+ Alpha metrics will not be analyzed as they don't have any stability guarantees.
+ Their declaration will not have any restrictions, thus allowing dynamic generation.
+
+### Discovery of golang source files
+
+Stable metric list will be a bazel genrule provided with locations of files from `//:all-src` target.
+This target is auto-managed, validate in CI by `./hack/validate-bazel.sh` and should include all files in repository.
+ Golang source files will be filtered by extension.
+
+List of skipped directories:
+* `vendor/` - Kubernetes metrics from external repos will be shared metrics that will be defined in k/k and injected as
+              dependency during registration
+
+### Format of defining stable metrics
+
+To simplify static analysis implementation and reduce chance of missing metrics, we will restrict how stable metrics can be defined.
+Stable metrics will use exactly the same functions as normal ones, but code defining them will need to comply to the specific format:
+
+* Metric options (`CounterOpts`) needs to be directly passed to function (`kubemetrics.NewCounterVec`).
```

Comment:
These are just examples so we should makes sure to mark them as such in this text

Reply:
Added "e.g.". Thanks